File: platform/wa-connector/src/api-client.ts
import { config } from "./config.js";

export type WaSessionInfo = {
  id: string;
  org_id: string;
  label: string;
  external_id: string;
  wa_jid: string;
  pairing_state: string;
  status: string;
};

export type AuthState = {
  account_id: string;
  creds_json: string;
  keys_json: string;
};

export type ClaimedJob = {
  id: string;
  account_id: string;
  lead_id: string | null;
  target_name: string;
  target_jid: string;
  body: string;
  sender_type: string;
  status: string;
  trace_id?: string;
};

async function request<T>(
  method: string,
  path: string,
  body?: unknown
): Promise<T> {
  const res = await fetch(`${config.apiBaseUrl}${path}`, {
    method,
    headers: {
      "Content-Type": "application/json",
      "X-Connector-Key": config.connectorKey,
    },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`${method} ${path} → ${res.status} ${text.slice(0, 300)}`);
  }
  if (res.status === 204) return undefined as T;
  return (await res.json()) as T;
}

export const api = {
  listSessions: () => request<WaSessionInfo[]>("GET", "/internal/wa/sessions"),
  getAuth: (accountId: string) =>
    request<AuthState>("GET", `/internal/wa/sessions/${accountId}/auth`),
  putAuth: (accountId: string, creds_json: string, keys_json: string) =>
    request<AuthState>("PUT", `/internal/wa/sessions/${accountId}/auth`, {
      creds_json,
      keys_json,
    }),
  clearAuth: (accountId: string) =>
    request<{ ok: boolean }>("DELETE", `/internal/wa/sessions/${accountId}/auth`),
  putPairState: (
    accountId: string,
    payload: {
      pairing_state: string;
      qr_payload?: string;
      wa_jid?: string;
      status?: string;
      external_id?: string;
    }
  ) => request("PUT", `/internal/wa/sessions/${accountId}/pair-state`, payload),
  heartbeat: (accountId: string) =>
    request<{ ok: boolean }>("POST", `/internal/wa/sessions/${accountId}/heartbeat`),
  ingest: (accountId: string, body: Record<string, unknown>) =>
    request("POST", `/internal/wa/sessions/${accountId}/ingest`, body),
  claimJobs: (accountId: string, limit = 5) =>
    request<{ jobs: ClaimedJob[] }>(
      "POST",
      `/internal/wa/jobs/claim?account_id=${encodeURIComponent(accountId)}&limit=${limit}`
    ),
  completeJob: (jobId: string, ok: boolean, error = "") =>
    request(
      "POST",
      `/internal/wa/jobs/${jobId}/complete?ok=${ok ? "true" : "false"}&error=${encodeURIComponent(error)}`
    ),
  getPairCommand: (accountId: string) =>
    request<{ account_id: string; pairing_state: string; status: string; wa_jid: string }>(
      "GET",
      `/internal/wa/sessions/${accountId}/pair-command`
    ),
};

File: platform/bale-connector/api_client.py
from __future__ import annotations

from typing import Any

import requests

from config import API_BASE, CONNECTOR_KEY


class ApiClient:
    def __init__(self) -> None:
        self._s = requests.Session()
        self._s.headers.update(
            {
                "X-Connector-Key": CONNECTOR_KEY,
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        )

    def _url(self, path: str) -> str:
        return f"{API_BASE}{path}"

    def list_sessions(self) -> list[dict[str, Any]]:
        r = self._s.get(self._url("/internal/bale/sessions"), timeout=20)
        r.raise_for_status()
        data = r.json()
        return data if isinstance(data, list) else []

    def get_auth(self, account_id: str) -> dict[str, Any]:
        r = self._s.get(self._url(f"/internal/bale/sessions/{account_id}/auth"), timeout=20)
        r.raise_for_status()
        return r.json()

    def put_cursors(self, account_id: str, cursors: dict[str, Any]) -> None:
        r = self._s.put(
            self._url(f"/internal/bale/sessions/{account_id}/cursors"),
            json=cursors,
            timeout=20,
        )
        r.raise_for_status()

    def put_pair_state(
        self,
        account_id: str,
        *,
        pairing_state: str = "",
        status: str = "",
        external_id: str = "",
        label: str = "",
    ) -> None:
        r = self._s.put(
            self._url(f"/internal/bale/sessions/{account_id}/pair-state"),
            params={
                "pairing_state": pairing_state,
                "status": status,
                "external_id": external_id,
                "label": label,
            },
            timeout=20,
        )
        r.raise_for_status()

    def heartbeat(self, account_id: str) -> None:
        r = self._s.post(self._url(f"/internal/bale/sessions/{account_id}/heartbeat"), timeout=15)
        r.raise_for_status()

    def ingest(self, account_id: str, payload: dict[str, Any]) -> dict[str, Any]:
        r = self._s.post(
            self._url(f"/internal/bale/sessions/{account_id}/ingest"),
            json=payload,
            timeout=45,
        )
        r.raise_for_status()
        return r.json()

    def claim_jobs(self, account_id: str, limit: int = 5) -> list[dict[str, Any]]:
        r = self._s.post(
            self._url("/internal/bale/jobs/claim"),
            params={"account_id": account_id, "limit": limit},
            timeout=20,
        )
        r.raise_for_status()
        return list((r.json() or {}).get("jobs") or [])

    def complete_job(self, job_id: str, *, ok: bool = True, error: str = "") -> None:
        r = self._s.post(
            self._url(f"/internal/bale/jobs/{job_id}/complete"),
            params={"ok": str(ok).lower(), "error": error or ""},
            timeout=20,
        )
        r.raise_for_status()


api = ApiClient()
